### DISCUSSION TITLE
ICC Conversion Adds Extra Bands to Image

### DISCUSSION THREAD

**Discussion by turtletowerz** [libvips/libvips#3361] [Help wanted]
I'm not sure if this is a bug or intended functionality so I wanted to get some feedback here before opening a proper issue.

When attempting to convert [this](https://cdn.donmai.us/original/f5/7c/f57ce27db2de478dea3106a61a1a3514.jpg) image's ICC profile to sRGB it outputs an image with 5 bands. Any operation after this (colorspace conversion and resizing have been tested) maintain the incorrect channel count. This doesn't seem to affect PNG/JPEG image writing but when writing to memory it results in a bad number of channels. Example code:
```c
#include <vips/vips.h>

void *modifyImage(const char *path, size_t *size, double width, double height) {
	VipsImage *im, *x;
	void *data;

	if (!(im = vips_image_new_from_file(path, "access", VIPS_ACCESS_SEQUENTIAL, NULL))) {
		return NULL;
	}

	g_log("vips-icc-test", G_LOG_LEVEL_INFO, "bands = %d\n", im->Bands);

	if (vips_image_get_typeof(im, VIPS_META_ICC_NAME)) {
		if (vips_icc_transform(im, &x, "srgb", NULL)) {
			return NULL;
		}
		g_object_unref(im);
		im = x;
	}
	
	g_log("vips-icc-test", G_LOG_LEVEL_INFO, "bands = %d\n", im->Bands);
	
	if (vips_image_get_interpretation(im) != VIPS_INTERPRETATION_sRGB) {
		if (vips_colourspace(im, &x, VIPS_INTERPRETATION_sRGB, NULL)) {
			return NULL;
		}
		g_object_unref(im);
		im = x;
	}

	g_log("vips-icc-test", G_LOG_LEVEL_INFO, "bands = %d\n", im->Bands);

	if (vips_resize(im, &x, width / (im->Xsize), "vscale", height / (im->Ysize), "kernel", VIPS_KERNEL_LINEAR, NULL)) {
		return NULL;
	}
	g_object_unref(im);
	im = x;

	if (!vips_image_hasalpha(im)) {
		if (vips_addalpha(im, &x, NULL)) {
			return NULL;
		}
		g_object_unref(im);
		im = x;
	}

	if (!(data = vips_image_write_to_memory(im, size))) {
		return NULL;
	}

	g_object_unref(im);
	return data;
}
```
Output
```
bands = 3
bands = 5
bands = 5
```

A small set of images have been tested and it seems to be Greyscale ICC profiles on images with an sRGB (or non-greyscale) colorspace. Is this intended functionality intended and if so how should I go about dealing with outputs that may have an arbitrary number of color bands?

**Comment by jcupitt** [comment]
Hi @turtletowerz,

> A small set of images have been tested and it seems to be Greyscale ICC profiles on images with an sRGB (or non-greyscale) colorspace. Is this intended functionality intended and if so how should I go about dealing with outputs that may have an arbitrary number of color bands?

That's right, if libvips sees a three band image plus a monochrome ICC profile it assumes the three band image is GAA, ie. a mono image plus two alpha channels. When you convert to sRGB you get RGBAA, ie. the first G is used for the profile and the two alphas are copied over.

You can see this behaviour most clearly with eg.:

```
$ vips icc_import ~/pics/f57ce27db2de478dea3106a61a1a3514.jpg  x.v
$ vipsheader x.v
x.v: 2362x2362 float, 5 bands, lab, jpegload
```

So importing your image to PCS generates LABAA.

A few people have been bitten by this now :( and I agree it's not obvious behaviour.

At the moment when the image and the profile are in conflict, libvips fixes it by adjusting the image interpretation, ie. it does the minimal change. Alternative behaviours might be:

1. Refuse to process images where the profile is incorrect.
2. Resolve the conflict by adding an extra colourspace conversion, so in this case insert an extra RGB->mono before the import. In other cases of bad profiles, you might need to insert mono->CMYK or something even more difficult.
3. Keep the current behaviour where conflicts are resolved where possible by assuming extra bands are alpha.

I think 1. would probably be the safest, but would be annoying. It would force people to add extra code to handle incorrectly tagged images, which might be a good thing overall. Option 2. is superficially attractive, but I think it would not be possible to make libvips resolve all conflicts correctly.

**Comment by jcupitt** [comment]
I would use `thumbnail` instead of open/colourspace/resize --- it should be quite a bit quicker, at least in some cases.

**Comment by turtletowerz** [comment]
Thanks for the quick feedback @jcupitt !

I agree that the alternatives are undesirable and it isn't really something vips should be dealing with since it's case-by-case. ~~On a dozen or so images I tested the alpha channels seemed to be duplicated, do you think it would be safe to do something like~~
I'm not convinced this is true anymore since when the profile is applied it should convert the R channel to greyscale

As for the vips_thumbnail suggestion, It does seem considerably faster, but I'm getting some weird results.
```c
if (vips_thumbnail(path, &im, 10, "height", 10, "export_profile", "srgb", "fail_on", VIPS_FAIL_ON_ERROR, NULL)) {
	return NULL;
}

g_log(DOMAIN, G_LOG_LEVEL_INFO, "%dx%d, bands = %d\n", vips_image_get_width(im), vips_image_get_height(im), im->Bands);
```
Edit: After testing a few more things the issue seemed to be that `VIPS_SIZE_FORCE` wasn't set. I found this un-intuitive, assuming that setting a width and a height would force the size anyways. Though it would break backwards compatibility, automatically setting `VIPS_SIZE_FORCE` if width and height are set (assuming no other size is set) would be a nice change.

**Reply by kleisauke** [reply]
`vips_thumbnail` resize behavior is based on a square bounding box, the future possible enhancement at #1639 or #709 could improve this (but the former PR would break API compatibility). For now, you can pass `VIPS_MAX_COORD`  to prevent reduction or enlargement in that axis.
https://github.com/libvips/libvips/blob/65c99219af5b3f4a8e9d4a9daf7886a6c97cf619/tools/vipsthumbnail.c#L395-L399

Note that `VIPS_SIZE_FORCE` breaks aspect ratio, so it's usually not recommended.

**Reply by turtletowerz** [reply]
Yes I want to break aspect ratio because the image needs to be exactly the `width x height` dimensions I provide. I read the following from the docs

> The output image will fit within a square of size width x width . You can specify a separate height with the height option.

And assumed that meant it would be that exact size, which is admittedly a bad assumption to make.

**Comment by evazion** [comment]
I think the current behavior is a bug and a regression since 8.13. Generating a thumbnail of a color image with a monochrome profile now produces a monochrome thumbnail:

```
wget https://cdn.donmai.us/original/a6/bd/a6bd9d52b3ed9c09b908a525665b655e.jpg
vipsthumbnail a6bd9d52b3ed9c09b908a525665b655e.jpg -s 849x1000 -e srgb -o "output.jpg[strip]"
```

([input](https://cdn.donmai.us/original/a6/bd/a6bd9d52b3ed9c09b908a525665b655e.jpg), [output](https://cdn.donmai.us/original/91/9c/919c9181303e4380bb35b2ddcd3fb07c.jpg))

In 8.13 this would log a "profile incompatible with image" warning and ignore the profile. Most other programs ignore it too when a color image has a monochrome profile. It's not uncommon in the wild to find color JPEGs with a monochrome profile leftover from some earlier process.

I would argue that ignoring incompatible profiles and logging a warning is the correct behavior. The current behavior is difficult to deal with, since AFAICT there's no easy way in libvips for applications to check whether an image has a monochrome profile and decide whether to ignore it or apply it. There's `vips_icc_is_compatible_profile`, but it's not exposed in bindings and it returns true in this case anyway.

**Reply by kleisauke** [reply]
I opened PR #3372 for this.
